content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md

```diff
@@ -178,6 +192,273 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
+    type: string
+    pattern: "abc"
+  bar:
+    type: integer
+  metadata:
+    type: object
+    properties:
+      name:
+        type: string
+        pattern: "^a"
+anyOf:
+- properties:
+    bar:
+      minimum: 42
+  required: ["bar"]
+```
+
+Violations of the structural schema rules are reported in the `NonStructural` condition in the CustomResourceDefinition.
+
+Not being structural will disable the following features:
+
+* [Validation Schema Publishing](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#publish-validation-schema-in-openapi-v2)
+* [Webhook Conversion](/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion)
+* [Validation Schema Defaulting](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#defaulting)
+* [Pruning](#preserving-unknown-fields)
+
+and possibly more features in the future.
+
+### Pruning versus preserving unknown fields
+
+{{< feature-state state="beta" for_kubernetes_version="1.15" >}}
+
+CustomResourceDefinitions traditionally store any (possibly validated) JSON as is in etcd. This means that unspecified fields (if there is a [OpenAPI v3.0 validation schema](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) at all) will be persisted. This is in contrast to native Kubernetes resources like e.g. a pod. If the user specifies a field `spec.randomField`, this will be accepted by the API server HTTPS endpoint.
+
+If a [structural OpenAPI v3 validation schema](#specifying-a-structural-schema) is defined (either in the global `spec.validation.openAPIV3Schema` or for each version), pruning can be enabled which drops unspecified fields on creation and on update by setting `spec.preserveUnknownFields` to `false`.
+
+Compare the CustomResourceDefinition `crontabs.stable.example.com` above. It has pruning enabled. Hence, if you save the following YAML to `my-crontab.yaml`:
+
+```yaml
+apiVersion: "stable.example.com/v1"
+kind: CronTab
+metadata:
+  name: my-new-cron-object
+spec:
+  cronSpec: "* * * * */5"
+  image: my-awesome-cron-image
+  someRandomField: 42
+```
+
+and create it:
+
+```shell
+kubectl create -f my-crontab.yaml -o yaml
+```
+
+you should get:
+
+```console
+apiVersion: stable.example.com/v1
+kind: CronTab
+metadata:
+  creationTimestamp: 2017-05-31T12:56:35Z
+  generation: 1
+  name: my-new-cron-object
+  namespace: default
+  resourceVersion: "285"
+  selfLink: /apis/stable.example.com/v1/namespaces/default/crontabs/my-new-cron-object
+  uid: 9423255b-4600-11e7-af6a-28d2447dc82b
+spec:
+  cronSpec: '* * * * */5'
+  image: my-awesome-cron-image
+```
+
+The field `someRandomField` has been pruned.
+
+In `apiextensions/v1beta1`, pruning is disabled by default, i.e. `spec.preserveUnknownFields` defaults to `true`. In `apiextensions/v1` no new CustomResourceDefinitions with `spec.preserveUnknownFields: true` will be allowed to be created.
+
+### Controlling pruning
```

Comment:
nit: this is not specific to this PR

I wish the table of content looks like
```
- Specifying a structural schema
  - structural schema
    # group these three together
  - customizing structural schema with openapi extensions
    - Controlling pruning
    - IntOrString
    - RawExtension
- Advanced topics
  - Validation
  - Defaulting
    # I feel pruning doesn't have to be under structural schema
  - Pruning
```
but currently it only shows top level: https://deploy-preview-14583--kubernetes-io-vnext-staging.netlify.com/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#advanced-topics
```
- Specifying a structural schema
- Advanced topics
```

Reply:
I place pruning more at the top because this won't be opt-in anymore in 1.0.